### PR TITLE
[FINAL] Subscriber attributes

### PR DESCRIPTION
**Pull request by vegaro**
- Adds `invalidatePurchaserInfoCache`
- Adds subscriber attributes functions
- Adds another `addAttributionData` function that accepts a dictionary

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 2.1.0
+
+- Adds Subscriber Attributes, which allow developers to store additional, structured information
+for a user in RevenueCat. More info: https://docs.revenuecat.com/docs/user-attributes.
+
```
Comment (vegaro) [inline CHANGELOG.md:5]: Add reference to invalidate Purchaser info cache

File: RevenueCat/Scripts/Purchases.cs [hunk 2]
```diff
@@ -927,27 +1016,44 @@ public class EntitlementInfoResponse
...
+        public float price;
+        public string price_string;
+        [CanBeNull] public string currency_code;
+        public float intro_price;
+        public string intro_price_string;
+        public string intro_price_period;
+        public string intro_price_period_unit;
+        [CanBeNull] public int intro_price_period_number_of_units;
+        [CanBeNull] public int intro_price_cycles;
+    }
```
Comment (vegaro) [inline RevenueCat/Scripts/Purchases.cs:1058]: When this PR is merged, I would like to move all these classes to their own class.

File: RevenueCat/Scripts/PurchasesWrapperAndroid.cs [hunk 3]
```diff
@@ -181,5 +182,70 @@ public void CheckTrialOrIntroductoryPriceEligibility(string[] productIdentifiers
...
+
+        public DictionaryRequest(Dictionary<string, string> dictionary) {
+            keys = new string[dictionary.Keys.Count];
+            values = new string[dictionary.Values.Count];
+        
+            dictionary.Keys.CopyTo(keys, 0);
+            dictionary.Values.CopyTo(values, 0);
+        }
+
+    }
```
Comment (vegaro) [inline RevenueCat/Scripts/PurchasesWrapperAndroid.cs:248]: How can we extract this to be used in the WrapperIOS too? I don't want to make it public and expose it.

File: Android/purchasesunity/src/main/java/com/revenuecat/purchasesunity/PurchasesWrapper.java [hunk 4]
```diff
@@ -238,6 +239,41 @@ public static void checkTrialOrIntroductoryPriceEligibility(String jsonProducts)
 
     }
 
+    public static void invalidatePurchaserInfoCache() {
+        CommonKt.invalidatePurchaserInfoCache();
+    }
+
+    public static void setAttributes(String jsonAttributes) {
```
Comment (aboedo) [inline Android/purchasesunity/src/main/java/com/revenuecat/purchasesunity/PurchasesWrapper.java:246]: I'm a bit confused by this - if we're already sending json and parsing keys, why are we sending keys and values separately? Why not just parse the JSON into dict directly?

File: CHANGELOG.md (hunk 1 above)
Comment (aboedo) [reply CHANGELOG.md:5]: maybe we can start doing the link to release changelogs thing?

File: RevenueCat/Example/PurchasesListener.cs [hunk 5]
```diff
@@ -42,6 +43,19 @@ private void Start()
...
             }
         });
+
+        purchases.InvalidatePurchaserInfoCache();
+
+        var map = new Dictionary<string, string>();
+        map.Add("cat", "orange");
+        map.Add("dog", "brown");
+        map.Add("perrito", "");
+        map.Add("gatitio", null);
```
Comment (aboedo) [inline RevenueCat/Example/PurchasesListener.cs:53]: we don't get dogs or cats in Spanish?

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 6]
```diff
@@ -203,6 +208,36 @@ - (void)checkTrialOrIntroductoryPriceEligibility:(NSArray *)productIdentifiers
     }];
 }
 
+- (void)invalidatePurchaserInfoCache
+{
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:212]: there's some inconsistency in how we're doing formatting of `{`. 
In some methods, we're doing it in the line _after_ the signature, and in others we're doing it on the same line. 
same line seems to be the most prevalent in purchases-ios, so I'd go with that. 

No need to update the entire codebase for it, but we should pick a standard and apply it consistently from now on.

File: RevenueCat/Plugins/iOS/RCPurchases+HybridAdditions.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+#import <Foundation/Foundation.h>
+#import "RCPurchases.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+
+@interface RCPurchases (HybridAdditions)
+
+- (void)_setPushTokenString:(nullable NSString *)pushToken;
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/RCPurchases+HybridAdditions.h:14]: should this be a part of common, since it's used by more than one hybrid sdk?

File: RevenueCat/Scripts/Purchases.cs [hunk 8]
```diff
@@ -171,113 +173,104 @@ public bool IsAnonymous()
...
-    {
-        private readonly PurchaserInfoResponse _response;
-
-        public PurchaserInfo(PurchaserInfoResponse response)
-        {
-            _response = response;
         }
 
-        public EntitlementInfos Entitlements
+        public void InvalidatePurchaserInfoCache()
```
Comment (aboedo) [inline RevenueCat/Scripts/Purchases.cs:179]: does the noop wrapper _have_ to live in this file? the others live in separate files, and moving it would reduce noise on this file. 
Since there's already a defined interface, the noop wrapper doesn't add much when reading

File: RevenueCat/Scripts/Purchases.cs (hunk 2 above)
Comment (aboedo) [reply RevenueCat/Scripts/Purchases.cs:972]: +100

File: RevenueCat/Scripts/PurchasesWrapper.cs [hunk 9]
```diff
@@ -21,4 +22,10 @@ public interface IPurchasesWrapper
 	void SetAutomaticAppleSearchAdsAttributionCollection(bool enabled);
     bool IsAnonymous();
     void CheckTrialOrIntroductoryPriceEligibility(string[] productIdentifiers);
+    void InvalidatePurchaserInfoCache();
+    void SetAttributes(Dictionary<string, string> attributes );
```
Comment (aboedo) [inline RevenueCat/Scripts/PurchasesWrapper.cs:26]: extra space at end of line before `);`

**Comment by aboedo** [comment]
looking great so far, left a couple of comments

File: Android/purchasesunity/src/main/java/com/revenuecat/purchases/common/mappers.kt [hunk 10]
```diff
@@ -92,63 +102,130 @@ private fun Package.map(offeringIdentifier: String): Map<String, Any?> =
...
+                ?: 0)
+        ) + introductoryPricePeriod.mapPeriodDeprecated()
+    } else {
+        mapOf(
+            "intro_price" to null,
+            "intro_price_string" to null,
+            "intro_price_period" to null,
+            "intro_price_cycles" to null,
+            "intro_price_period_unit" to null,
+            "intro_price_period_number_of_units" to null
```
[the remark this answers is not in the document]
Comment (vegaro) [reply Android/purchasesunity/src/main/java/com/revenuecat/purchases/common/mappers.kt:130]: What about `introductoryPricePeriod.mapPeriodDeprecated()` ? `introductoryPricePeriod` is going to be null or blank so the result is the same

[the remark this answers is not in the document]
Comment (aboedo) [reply Android/purchasesunity/src/main/java/com/revenuecat/purchases/common/mappers.kt:130]: SGTM

File: RevenueCat/Plugins/iOS/RCPurchases+HybridAdditions.h (hunk 7 above)
Comment (vegaro) [reply RevenueCat/Plugins/iOS/RCPurchases+HybridAdditions.h:14]: This is part of common actually. We don't have the `.a` with all these files in Unity yet

File: RevenueCat/Scripts/Purchases.cs (hunk 8 above)
Comment (vegaro) [reply RevenueCat/Scripts/Purchases.cs:179]: No. It's part of the class extraction plan. I didn't want to do it in this PR because it would be hard to see the changes

File: Android/purchasesunity/local.properties [hunk 11]
```diff
@@ -0,0 +1,8 @@
+## This file must *NOT* be checked into Version Control Systems,
```
Comment (aboedo) [inline Android/purchasesunity/local.properties:1]: oops

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,25 @@
...
+ - [Release Notes 3.1.0](https://github.com/RevenueCat/purchases-ios/releases/tag/3.1.0)
+ - [Release Notes 3.0.4](https://github.com/RevenueCat/purchases-ios/releases/tag/3.0.4)
+ - [Release Notes 3.0.3](https://github.com/RevenueCat/purchases-ios/releases/tag/3.0.3)
+ - [Release Notes 3.0.2](https://github.com/RevenueCat/purchases-ios/releases/tag/3.0.2)
+
+**Android Native SDK Update 3.1.0**
+ - [Release Notes 3.1.0](https://github.com/RevenueCat/purchases-ios/releases/tag/3.1.0)
+ - [Release Notes 3.0.7](https://github.com/RevenueCat/purchases-ios/releases/tag/3.0.7)
+ - [Release Notes 3.0.6](https://github.com/RevenueCat/purchases-ios/releases/tag/3.0.6)
+ - [Release Notes 3.0.5](https://github.com/RevenueCat/purchases-ios/releases/tag/3.0.5)
```
Comment (aboedo) [inline CHANGELOG.md:21]: 👏

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 13]
```diff
@@ -182,18 +184,15 @@ - (void)purchases:(RCPurchases *)purchases didReceiveUpdatedPurchaserInfo:(RCPur
...
-- (BOOL)isAnonymous
-{
+- (BOOL)isAnonymous {
     return @([RCCommonFunctionality isAnonymous]);
 }
 
-- (void)checkTrialOrIntroductoryPriceEligibility:(NSArray *)productIdentifiers
-{
+- (void)checkTrialOrIntroductoryPriceEligibility:(NSArray *)productIdentifiers {
     [RCCommonFunctionality checkTrialOrIntroductoryPriceEligibility:productIdentifiers completionBlock:^(NSDictionary<NSString *,NSDictionary *> * _Nonnull responseDictionary) {
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:196]: break line before `completionBlock` so it's easier to read

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 14]
```diff
@@ -339,3 +361,35 @@ void _RCCheckTrialOrIntroductoryPriceEligibility(const char *productIdentifiersJ
...
     [_RCUnityHelperShared() checkTrialOrIntroductoryPriceEligibility:productsRequest[@"productIdentifiers"]];
 }
+
+void _RCInvalidatePurchaserInfoCache() {
+    [_RCUnityHelperShared() invalidatePurchaserInfoCache];
+}
+
+void _RCSetAttributes(const char* attributesJSON) {
+    NSError *error = nil;
+    NSDictionary *attributes = [NSJSONSerialization JSONObjectWithData:[convertCString(attributesJSON) dataUsingEncoding:NSUTF8StringEncoding] options:0 error:&error];
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:371]: we don't have a strict char count standard yet, but I think we should try to keep lines ideally under 120 where possible

Comment (aboedo) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:371]: one thought for this one would be 
```objective-c
NSData *attributesAsData = [convertCString(attributesJSON) dataUsingEncoding:NSUTF8StringEncoding];
NSDictionary *attributes = [NSJSONSerialization JSONObjectWithData:attributesAsData options:0 error:&error];
```
or even 

```objective-c
NSData *attributesAsData = [convertCString(attributesJSON) dataUsingEncoding:NSUTF8StringEncoding];
NSDictionary *attributes = [NSJSONSerialization JSONObjectWithData:attributesAsData 
                                                           options:0 
                                                             error:&error];
```

File: RevenueCat/Plugins/iOS/RCCommonFunctionality.m [hunk 15]
```diff
@@ -181,31 +224,63 @@ + (void)makeDeferredPurchase:(RCDeferredPromotionalPurchaseBlock)deferredPurchas
...
+                if (discountToUse) {
+                    [RCPurchases.sharedPurchases paymentDiscountForProductDiscount:discountToUse product:aProduct completion:^(SKPaymentDiscount *paymentDiscount, NSError *error) {
+                        if (paymentDiscount) {
+                            self.discounts[[paymentDiscount.timestamp stringValue]] = paymentDiscount;
+                            completion(paymentDiscount.dictionary, nil);
+                        } else {
+                            completion(nil, [self payloadForError:error  withExtraPayload:@{}]);
+                        }
+                    }];
+                } else {
```
Comment (aboedo) [inline RevenueCat/Plugins/iOS/RCCommonFunctionality.m:271]: [not needed for this PR] I kinda feel like these are complex enough that they might merit being in their own class. 

`RCCommonFunctionality` is right now acting as both a Facade and implementing logic for products & discounts. The latter feels a bit out of scope.

File: RevenueCat/Plugins/iOS/RCPurchases+HybridAdditions.h (hunk 7 above)
Comment (aboedo) [reply RevenueCat/Plugins/iOS/RCPurchases+HybridAdditions.h:14]: my bad, I was looking for the word "common" in the path, I forgot they're just loose files in Unity.

File: RevenueCat/Scripts/Purchases.cs [hunk 16]
```diff
@@ -390,7 +423,7 @@ public Package(PackageResponse response)
         {
             Identifier = response.identifier;
             PackageType = response.packageType;
-            Product = response.product;
+            Product = new Product(response.product);
```
Comment (aboedo) [inline RevenueCat/Scripts/Purchases.cs:426]: was this a bug?

File: RevenueCat/Scripts/Purchases.cs [hunk 17]
```diff
@@ -662,7 +737,12 @@ private void _receiveProducts(string productsJson)
         }
         else
         {
-            ProductsCallback(response.products, null);
+            var products = new List<Product>();
+            foreach (var productResponse in response.products)
+            {
+                products.Add(new Product(productResponse));
+            }
```
Comment (aboedo) [inline RevenueCat/Scripts/Purchases.cs:744]: I think you could replace these lines with
```c#
var products = response.products.Select(productResponse => new Product(productResponse))
```
https://blog.submain.com/csharp-functional-programming/

File: RevenueCat/Scripts/PurchasesWrapperAndroid.cs [hunk 18]
```diff
@@ -181,5 +183,53 @@ public void CheckTrialOrIntroductoryPriceEligibility(string[] productIdentifiers
         }
     }
 
+    public void InvalidatePurchaserInfoCache()
+    {
+        using (var purchases = new AndroidJavaClass("com.revenuecat.purchasesunity.PurchasesWrapper"))
+        {
+            purchases.CallStatic("invalidatePurchaserInfoCache");
```
Comment (aboedo) [inline RevenueCat/Scripts/PurchasesWrapperAndroid.cs:190]: [not for this PR] I wonder if c# has something akin to python's decorators, we could turn these into one-liners. even macros might help. 
¯\\_(ツ)_/¯

File: RevenueCat/Scripts/Purchases.cs (hunk 17 above)
Comment (vegaro) [reply RevenueCat/Scripts/Purchases.cs:744]: I saw that, but I was scared to change it just in case it's not compatible with all runtimes. I will double check

Comment (vegaro) [reply RevenueCat/Scripts/Purchases.cs:744]: https://forum.unity.com/threads/is-c-linq-fully-reliable-supported.379326/ I am not sure

File: RevenueCat/Scripts/Purchases.cs (hunk 16 above)
Comment (vegaro) [reply RevenueCat/Scripts/Purchases.cs:426]: no, it's because before `response.product` was a `Product`, now it's a `ProductResponse`

File: RevenueCat/Scripts/PurchasesWrapperAndroid.cs (hunk 18 above)
Comment (vegaro) [reply RevenueCat/Scripts/PurchasesWrapperAndroid.cs:190]: Now that you say, this could be a method haha

File: RevenueCat/Scripts/PurchasesWrapperAndroid.cs [hunk 19]
```diff
@@ -177,57 +126,54 @@ public void CheckTrialOrIntroductoryPriceEligibility(string[] productIdentifiers
...
+    }
+
+    private const string PurchasesWrapper = "com.revenuecat.purchasesunity.PurchasesWrapper";
+
+    private static void CallPurchases(string methodName, params object[] args)
+    {
+        using (var purchases = new AndroidJavaClass(PurchasesWrapper))
+        {
+            purchases.CallStatic(methodName, args);
+        }
```
Comment (aboedo) [inline RevenueCat/Scripts/PurchasesWrapperAndroid.cs:169]: nice!

File: RevenueCat/Scripts/PurchasesWrapperAndroid.cs (hunk 3 above)
Comment (aboedo) [reply RevenueCat/Scripts/PurchasesWrapperAndroid.cs:178]: does this create its own assembly? if so, moving it to an `internal` class should do the trick, right?

Comment (vegaro) [reply RevenueCat/Scripts/PurchasesWrapperAndroid.cs:178]: I don't remember what this was. What was it?

Comment (aboedo) [reply RevenueCat/Scripts/PurchasesWrapperAndroid.cs:178]: I believe it was a private class

Comment (aboedo) [reply RevenueCat/Scripts/PurchasesWrapperAndroid.cs:178]: or maybe a private method
